### PR TITLE
OrtResult: Associate package curations with provider IDs

### LINKED ISSUE #5668: Record where curations came from
Analyzer results that contain curations should record the provider a curation originates from. Maybe just the provider's (class) name would be enough, but e.g. for `CompositePackageCurationProvider` the real underlying provider should be recorded, and e.g. for `FilePackageCurationProvider` maybe some more details like the (relative) directory path would make sense.

Recording the origin / source of a curation improves transparency and helps with debugging metadata issues.

### PR DESCRIPTION
**Pull request by fviernau**
Keep the association between curations and providers (IDs) in the `OrtResult` to enable

1. Tracability of curations back to providers
2. supporting additional use cases
    - replace curations for specific provider with given ones
    - re-request curations for given provider IDs

See individual commits.

BREAKING CHANGE: No backwards compatibility for de-serializing ORT files. This is ok as backwards compatibility has been broken very recently, so it's not a big deal to extend this by a couple of days. The file format change is limited to `OrtResult.resolvedConfiguration`.

### CODE REVIEW

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (sschuberth) [inline cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: Without having given it too much thought yet, I rather would have expected a structure like the following:

```
resolved_configuration:
  package_curations:
  - provider_id: "DefaultFile"
    curations:
    - id: "Maven:org.hamcrest::"
      curations:
        comment: "Use the actual homepage instead of the GitHub page."
        homepage_url: "http://hamcrest.org/JavaHamcrest/"
    - id: "Maven:org.hamcrest:hamcrest-core:"
      curations:
        comment: "Fix description."
        description: "Curated description."
```

File: model/src/main/kotlin/OrtResult.kt
Comment (sschuberth) [inline model/src/main/kotlin/OrtResult.kt:331]: How about changing the signature of `replacePackageCurations` to take a `PackageCurationProvider` and an `id` string instead?

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (sschuberth) [inline model/src/main/kotlin/ResolvedConfiguration.kt:88]: Why is this still necessary with all the `distinct()` calls we now have in the provider implementations? Shouldn't we only have either-or?

**Review by sschuberth — DISMISSED** [review]

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (fviernau) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: Hm, @mnonnenmacher, @MarcelBochtler, you an me discussed this in a recent developer meeting and what we came up with is this: https://github.com/oss-review-toolkit/ort/issues/6188#issuecomment-1383833350. (That's also mentioned in the commit message)

So, this PR is targeting this proposal, but leaving out the provider meta-data for now.

So, can we stick to it as-is?

File: model/src/main/kotlin/OrtResult.kt
Comment (fviernau) [reply model/src/main/kotlin/OrtResult.kt:331]: I had a long fight with this particular thing, and it was IMO by far not trivial.
The callers are CLI commands. If your  proposal was implemented this would become a functional change of the CLI.

I would like to keep the CLI changes out of scope of this PR. 
I'm already working on how to best integrate the provider IDs into CLI, but this IMO is a bigger topic and deserves it's own PR.

Fine?

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: I decided for the distinct calls because IMO it just encodes the appropriate way of serializing an instance of this class. I believe it's a good thing to not introduce such assumption, considering also programatical use or (future) use from code paths we currently don't have in mind.

Does it make sense? (If not I'll ditch my preference for you)

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (sschuberth) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: I'd be curious about the rationale that you guys discussed for the redundancy of listing the provider ids both inside `providers` *and* below `data`. Maybe I'm missing something, but this seems overly complicated to me.

Comment (sschuberth) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: Ah, it looks so redundant because the providers in this example do not require any configuration. I guess the idea was to make `providers` inside `resolved_configuration` more or less a copy of `packageCurationProviders` from `config.yml`, right?

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (sschuberth) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: IMO we should have `distinct()` only in one place, and preferably in a place what does not require / rely on all curation provider implementers to remember calling `distinct()`. So I guess we should have it here, but not in the implementations. Or extend the `PackageCurationProvider` with a public and an internal function, where the public one just calls the internal one plus `distinct()`, and implementers only implement the internal function.

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (fviernau) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: > I'd be curious about the rationale that you guys discussed for the redundancy of listing the provider ids both inside providers and below data. Maybe I'm missing something, but this seems overly complicated to me.

The proposal was created in a discussion with you @sschuberth, @mnonnenmacher , @MarcelBochtler and me. 
That model wasn't, as far as I remember discussed any further without you being present. 

> Ah, it looks so redundant because the providers in this example do not require any configuration. I guess the idea was to make providers inside resolved_configuration more or less a copy of packageCurationProviders from config.yml, right?

Yes.

File: model/src/main/kotlin/OrtResult.kt
Comment (fviernau) [reply model/src/main/kotlin/OrtResult.kt:331]: > I don't think so.

Well, that dependes on the semantic you envision for ` replacePackageCurations(provider, id = "EvaluatorCommandOption")`. Could you explain this?

Anyhow, I believe it is fine to consider this out of scope of this commit,
and I still strongly prefer to draw the line at this TODO comment and let me address it in my next PR.

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: From my perspective the `distinct()` call should be present in both, (1) the providers and (2) the `ResolvedConfiguration` serialization. I argue that both components can be used independently from one another and considering both in isolation it makes sense to me that 1. providers return distinct configurations and 2. serialization drops duplicates. I'm also fine with just disagreeing on that one.

I'm not sure what exactly you have in mind with "public and an internal function", but I have a hunch that this could again be out of scope / separate commit if at all.

Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: I still fear this is some kind of mutual misunderstanding involved in the discussion, maybe let's have a call to shortcut this.

Comment (sschuberth) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: I agree having a call would help. Maybe we could even shift the developer meeting from next Monday to tomorrow already, as on Monday I'll probably not be available.

Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: That'd be awesome. I'll propose it tomorrow morning in the group chat.

File: model/src/main/kotlin/OrtResult.kt
Comment (fviernau) [reply model/src/main/kotlin/OrtResult.kt:331]: I've implemented your above proposal.

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:88]: The distinct calls have been eliminated as discussed:
- the provider API now returns sets of curations, done in a separate already merged PR.
- the `ResolvedConfiguration` now also holds sets instead of lists of curations.

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (fviernau) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: @sschuberth and me agreed on discussing this in the @oss-review-toolkit/core-devs round.

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (sschuberth) [inline model/src/main/kotlin/ResolvedConfiguration.kt:87]: > Sorts the keys of [PackageCurations.data] alphabetically

Should we really do this? Wouldn't it be more natural to simply use the same order as in `providers`?

> the values of [PackageCurations.data] by their [id][PackageCuration.id]

Similarly here, even if we agreed that the order of curations within a single provider should not matter, shouldn't we simply maintain the order in which the provider returned them? This would easy debugging if we ever ran into the case where the order of curations from a single provider *does* matter.

> drops entries for providers without curations from [PackageCurations.data].

IMO those should already be filtered out in `resolveConfiguration()`.

Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:87]: > Should we really do this? Wouldn't it be more natural to simply use the same order as in providers?

I believe there is value in having things ordered. In particular:

1. easier to find things
2. related things are next to on another

I'm not quite sure what you want to say with the term "more natural". Can you rephrase please?
Note that we agreed as part of our discussion on this PR that for a particular provider the order does not have meaning, so it is arbitrary. I don't see why arbitrary order is natural and better than sorted order.

Comment (sschuberth) [reply model/src/main/kotlin/ResolvedConfiguration.kt:87]: > for a particular provider the order does not have meaning

Right, but my sentence that you're quoting referred to "the keys of [PackageCurations.data]", i.e. the ids of the providers. I would expect the ids here to match the ids below `providers`. That what I refer to as "natural".

Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:87]: > Right, but my sentence that you're quoting referred to "the keys of [PackageCurations.data]", i.e. the ids of the providers. I would expect the ids here to match the ids below `providers`. That what I refer to as "natural".

Oh, I see. I've got not that strong preference for the key ordering, but slightly prefer to keep the order as-is. Let's just also add this to our core devs discussion on it as the PR is blocked until then anyway .

File: advisor/src/test/assets/ort-analyzer-result.yml
Comment (mnonnenmacher) [inline advisor/src/test/assets/ort-analyzer-result.yml:574]: Commit message: There is a difference between footnotes and bottom links. For links use `[1]`, for footnotes use `[^1]`. Also, there is a colon missing after `[^1]`.

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (mnonnenmacher) [inline model/src/main/kotlin/ResolvedConfiguration.kt:63]: Nit: Fits in a single line.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml
Comment (fviernau) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: Outcome of discussion with @sschuberth @mnonnenmacher @MarcelBochtler : Let's change to:
```
resolved_configuration:
  package_curations:
  - provider:
      id: "CustomFile"
      type: "File"
      config:
        path: '/some-path/curations-dir'
        mustExist: false
    curations:
    - id: "Maven:org.hamcrest::"
      curations:
        comment: "Use the actual homepage instead of the GitHub page."
        homepage_url: "http://hamcrest.org/JavaHamcrest/"
    - id: "Maven:org.hamcrest:hamcrest-core:"
      curations:
        comment: "Fix description."
        description: "Curated description."
```

1. provider highest prio first
2. list of curations is ordered like the providers provided

Comment (fviernau) [reply cli/src/funTest/assets/gradle-all-dependencies-expected-result-with-curations.yml:476]: Implementation changed according to above comment.

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:63]: Not applicable anymore.

File: advisor/src/test/assets/ort-analyzer-result.yml
Comment (mnonnenmacher) [reply advisor/src/test/assets/ort-analyzer-result.yml:574]: They should not be. For bottom links GitHub creates a link directly to the URL, for footnotes GitHub creates an anchor link so it just scrolls down a little bit to where the footnote is (actually it always scrolled too far for me).

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (mnonnenmacher) [inline model/src/main/kotlin/ResolvedConfiguration.kt:56]: I would propose `ResolvedPackageCurations` as name for the class, it matches the context where it is used.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/ResolvedConfiguration.kt
Comment (fviernau) [reply model/src/main/kotlin/ResolvedConfiguration.kt:56]: @mnonnenmacher  are you ok if I make this a commit on top ?

Comment (mnonnenmacher) [reply model/src/main/kotlin/ResolvedConfiguration.kt:56]: Yes

File: advisor/src/test/assets/ort-analyzer-result.yml
Comment (sschuberth) [inline advisor/src/test/assets/ort-analyzer-result.yml:578]: This is something that I believe we forgot to discuss in our meeting, but IMO providers that don't have any curations should not be listed here. This is about *resolved* (i.e. relevant) configuration after all, and configured providers that don't contribute anything are not relevant.

File: analyzer/src/main/kotlin/Analyzer.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/Analyzer.kt]: The `distinct()` should have actually been removed as part of 7ab57c2a2ca040bfd3b21542fb1b47d72c495af7. For clarity, could you please extract this to a preceding commit that is documented as a fixup for 7ab57c2a2ca040bfd3b21542fb1b47d72c495af7?

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by mnonnenmacher — DISMISSED** [review]

File: advisor/src/test/assets/ort-analyzer-result.yml
Comment (mnonnenmacher) [reply advisor/src/test/assets/ort-analyzer-result.yml:578]: On the other hand this empty entry documents that the provider was queried and did not provide any applicable curations.

Comment (fviernau) [reply advisor/src/test/assets/ort-analyzer-result.yml:578]: I consciously decided for listing all enabled providers, to have the priorities of the providers in the ort result.
This enables the use case for re-resolving curations for any particular provider ID, no matter if the initial resolution lead to an empty result. Therefore I have a rather strong preference to keep all enabled providers in the list.

ok with that @mnonnenmacher  @sschuberth ?

Comment (sschuberth) [reply advisor/src/test/assets/ort-analyzer-result.yml:578]: > On the other hand this empty entry documents that the provider was queried and did not provide any applicable curations.

But what does that tell you? For example, it doesn't tell you that you could disable the provider, as it might have applicable curations for a slightly different set of packages.

**Review by sschuberth — APPROVED** [review]

File: advisor/src/test/assets/ort-analyzer-result.yml
Comment (mnonnenmacher) [reply advisor/src/test/assets/ort-analyzer-result.yml:578]: > > On the other hand this empty entry documents that the provider was queried and did not provide any applicable curations.
> 
> But what does that tell you? For example, it doesn't tell you that you could disable the provider, as it might have applicable curations for a slightly different set of packages.

It tells you that at the time the ORT result was created the provider had no curations for the packages contained in the result.

Comment (sschuberth) [reply advisor/src/test/assets/ort-analyzer-result.yml:578]: > It tells you that at the time the ORT result was created the provider had no curations for the packages contained in the result.

Obviously, yes. But I have doubts that this is valuable information. Anyway, what convinced me more is the re-resolving use-case and to maintain the order of providers there.

**Pull request merged**
